File: 2023/java/d06p1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

public class d06p1 {
    private static final int ZERO = 48;

    public static void main(String[] args) throws IOException {
        System.out.println("Day 6 - Part 1");

        BufferedReader reader = new BufferedReader(new FileReader("../_input/day06.txt"));

        long total = process(reader);
        reader.close();

        System.out.println("Total: " + total);
    }

    private static long process(BufferedReader reader) throws IOException {
        String rawTimes = null;
        String rawDists = null;

        String line;
        while ((line = reader.readLine()) != null) {
            if (rawTimes == null) rawTimes = line;
            else rawDists = line;
        }

        long[] times = getNums(rawTimes.split(":")[1]);
        long[] distances = getNums(rawDists.split(":")[1]);

        long total = 1;
        for (int i = 0; i < times.length; i++) {
            var t = times[i];
            var d = distances[i];

            long counter = 0;
            for (int speed = 0; speed <= t; speed++) {
                var timeLeft = t - speed;
                var dis = timeLeft * speed;
                if (dis > d) counter++;
            }
            total *= counter;
        }

        return total;
    }

    private static long[] getNums(String str) {
        String[] raws = str.split(" ");
        ArrayList<Long> builder = new ArrayList<>();
        for (var r : raws) {
            r = r.trim();
            if (!r.isEmpty()) builder.add(stringToLong(r));
        }
        return builder.stream().mapToLong(i -> i).toArray();
    }

    private static long stringToLong(String str) {
        long acc = 0;
        for (int i = 0; i < str.length(); i++) {
            long c = str.charAt(i);
            acc += (long) (Math.pow(10.0, str.length() - i - 1) * (c - ZERO));
        }
        return acc;
    }
}

File: 2023/rust/src/bin/d06p1.rs
use std::fs;

fn main() {
    println!("Day 6 - Part 1");

    let str = fs::read_to_string("../_input/day06.txt").unwrap();
    let lines: Vec<&str> = str.lines().collect();

    let total = process(lines);
    println!("Total: {}", total);
}

fn process(lines: Vec<&str>) -> u32 {
    let times = process_line(lines[0]);
    let distance = process_line(lines[1]);

    let mut total = 1;
    for (t, d) in times.iter().zip(distance.iter()) {
        let mut counter = 0;
        for speed in 0..=*t {
            let time_left = t - speed;
            let dis = time_left * speed;
            if dis > *d {
                counter += 1;
            }
        }
        total *= counter;
    }

    total
}

fn process_line(line: &str) -> Vec<u32> {
    line.split(':').collect::<Vec<&str>>()[1]
        .split_whitespace()
        .map(str_to_u32)
        .collect()
}

fn str_to_u32(str: &str) -> u32 {
    let mut acc = 0;
    for (i, c) in str.chars().enumerate() {
        acc += (10_u32.pow((str.len() - i - 1) as u32)) * c.to_digit(10).unwrap();
    }
    acc
}

#[cfg(test)]
mod test {
    use crate::process;

    #[test]
    fn test() {
        assert_eq!(
            process(
                r"Time:      7  15   30
Distance:  9  40  200"
                    .split('\n')
                    .collect()
            ),
            288
        );
    }
}
